Implement `Abstract_File_Check` base class as well as two plugin requirement checks using it

felixarntz opened the pull request:
### Description of the Change

* Implements an `Abstract_File_Check` base class to make it easy to implement checks like those used by the WordPress theme check plugin or by the early plugin review tool, as described in #174
* Implement two essential checks for minimum plugin requirements that use the new base class:
    * `Code_Obfuscation_Check`, based on https://github.com/WordPress/plugin-check/blob/trunk/checks/code-obfuscation.php
    * `File_Type_Check`, based on https://github.com/WordPress/plugin-check/blob/trunk/checks/file-checks.php
* Test coverage for the new checks is added, also indirectly covering the base class.

Closes #174

### How to test the Change
* Install a plugin that contains any of the following:
    * compressed files, e.g. ZIP, TAR, GZ, etc.
    * `.phar` files
    * version control checkouts, e.g. a `.git` or `.svn` folder
    * hidden files starting with `.`
    * obfuscated PHP files obfuscated with either Zend Guard, Source Guardian, or ionCube
* Ensure those plugin rule violations are flagged when running the plugin checker for that plugin.
* Alternatively to real plugins, you can also install/activate the relevant plugins in the `tests/phpunit/testdata/plugins` folder.

### Checklist:
- [x] I agree to follow this project's [**Code of Conduct**](https://github.com/10up/.github/blob/trunk/CODE_OF_CONDUCT.md).
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my change.
- [x] All new and existing tests pass.

Commit:
Implement Abstract_File_Check base class as well as two plugin requirement checks using it.

Commit:
Rename method.

Commit:
Fix PHPCS violations.

Commit:
Fix typo.

Commit:
Simplify check for whether to capture a warning or error.

Commit:
Add tests for the two new checks.

Commit:
Fix incorrect class name in comment.

Commit:
Fix fatal error from test plugin in CI.

mukeshpanchal27 commented on includes/Checker/Checks/File_Type_Check.php:
```diff
@@ -0,0 +1,184 @@
+		$directories = array_flip( array( '.git', '.svn', '.hg', '.bzr' ) );
+
+		$vcs_directories = array_filter(
+			array_unique(
```
As we have static array above we don't get any duplicate item.

felixarntz replied:
This doesn't filter the above array though. It filters the list of files, and as soon as more than 1 file is in the same directory, there will be duplicates.

mukeshpanchal27 commented on tests/phpunit/testdata/plugins/test-plugin-code-obfuscation-ioncube-errors/load.php:
```diff
@@ -0,0 +1,17 @@
+ * Author URI: https://make.wordpress.org/performance/
+ * License: GPLv2 or later
+ * License URI: https://www.gnu.org/licenses/old-licenses/gpl-2.0.html
+ * Text Domain: test-plugin-code-obfuscation-ioncube-errors
```
Do we need to include `Requires at least`, `Requires PHP`, and `Version` for all new plugins?

spacedmonkey replied:
These are test plugins, I don't think they are needed?

felixarntz replied:
Agreed, no need to include those things as they are test plugins.

mukeshpanchal27 commented on tests/phpunit/testdata/plugins/test-plugin-code-obfuscation-sourceguardian-errors/obfuscated.php:
```diff
@@ -0,0 +1,4 @@
+<?php
+if(function_exists('sg_load')) {
```
```suggestion
if ( function_exists( 'sg_load' ) ) {
```
Correct format

felixarntz replied:
Even though this violates WPCS, this must not be changed since that is the string to look for to detect Source Guardian. See the original check linked in the PR description.

It's a test plugin anyway, which is excluded from linting for reasons like this.

mukeshpanchal27 commented on includes/Checker/Checks/Code_Obfuscation_Check.php:
```diff
@@ -0,0 +1,143 @@
+			$result->add_message(
+				true,
+				sprintf(
+					'Code Obfuscation tools are not permitted. Detected: %s',
```
Should use translation function for all this strings. WDYT?

felixarntz replied:
Good point, I copied that over from the other code. Will update to use translation functions.

mukeshpanchal27 reviewed (commented):
Great work, @felixarntz! 👏  I have left some feedback and have a few questions. Overall, it looks solid to me.

jjgrainger reviewed (commented):
Thanks @felixarntz all looks good, though I think some of @mukeshpanchal27 feedback may be worth addressing.

spacedmonkey commented on includes/Checker/Checks/Abstract_File_Check.php:
```diff
@@ -0,0 +1,227 @@
+		// Backward compatibility prior to PHP 8.
+		$find_needle = 'str_contains';
+		if ( ! function_exists( 'str_contains' ) ) {
+			$find_needle = function( $haystack, $needle ) {
+				return false !== strpos( $haystack, $needle );
+			};
```
Could we just require `wp-includes/compat.php` here?

felixarntz replied:
I forgot that we require WP 6.0+ for the plugin checker, therefore I added this for older WP version support. But not needed, I remove this entirely, since `str_contains()` should always exist in WP 6.0+.

spacedmonkey commented on includes/Checker/Checks/File_Type_Check.php:
```diff
@@ -0,0 +1,184 @@
+			foreach ( $compressed_files as $file ) {
+				$result->add_message(
+					true,
+					'Compressed files are not permitted.',
```
Should this not be translated?

felixarntz replied:
See https://github.com/10up/plugin-check/pull/175#discussion_r1205828773, will fix.

spacedmonkey commented on includes/Checker/Checks/File_Type_Check.php:
```diff
@@ -0,0 +1,184 @@
+			foreach ( $phar_files as $file ) {
+				$result->add_message(
+					true,
+					'Phar files are not permitted.',
```
Should this not be translated?

felixarntz replied:
See above.

spacedmonkey commented on includes/Checker/Checks/File_Type_Check.php:
```diff
@@ -0,0 +1,184 @@
+	 * @param array        $files  List of absolute file paths.
+	 */
+	protected function look_for_vcs_directories( Check_Result $result, array $files ) {
+		$directories = array_flip( array( '.git', '.svn', '.hg', '.bzr' ) );
```
Wonder if we should also add .idea, as phpstorm adds these as well.

felixarntz replied:
This test is only for version control directories, `.idea` is an unrelated hidden file, which will anyways be covered by the `look_for_hidden_files()` method.

spacedmonkey commented on includes/Checker/Checks/File_Type_Check.php:
```diff
@@ -0,0 +1,184 @@
+			foreach ( $vcs_directories as $dir ) {
+				$result->add_message(
+					$is_error,
+					'Version control checkouts should not be present.',
```
https://github.com/10up/plugin-check/pull/175/files#r1205362655

spacedmonkey commented on includes/Checker/Checks/File_Type_Check.php:
```diff
@@ -0,0 +1,184 @@
+	 * @param Check_Result $result The check result to amend, including the plugin context to check.
+	 * @param array        $files  List of absolute file paths.
+	 */
+	protected function look_for_vcs_directories( Check_Result $result, array $files ) {
```
Not, I have used composer to deploy WP + plugins to production. If composer does a git checkout, this would result in .git files in production.

felixarntz replied:
This is taken over from the plugin review team's tool, so I think we should stick with that for now for parity. If this becomes a concern, we can further down in the future report an issue, but let's not do that just yet.

spacedmonkey commented on includes/Checker/Checks/File_Type_Check.php:
```diff
@@ -0,0 +1,184 @@
+	 */
+	protected function check_files( Check_Result $result, array $files ) {
+		if ( $this->flags & self::TYPE_COMPRESSED ) {
+			$this->look_for_compressed_files( $result, $files );
```
These 4 methods seem like they do the same thing. Could we make this more generic like this. 
```
$this->look_for_files( $result, $files, array( 'zip', 'gz', 'tgz', 'rar', 'tar', '7z' ),  'Compressed files are not permitted.', 'compressed_files' );
```

felixarntz replied:
Will update.

spacedmonkey commented on includes/Checker/Checks/Code_Obfuscation_Check.php:
```diff
@@ -0,0 +1,143 @@
+			$result->add_message(
+				true,
+				sprintf(
+					'Code Obfuscation tools are not permitted. Detected: %s',
```
Should these be translated?

spacedmonkey commented on includes/Checker/Checks/Abstract_File_Check.php:
```diff
@@ -0,0 +1,227 @@
+	 * @param string $needle The substring to search for.
+	 * @return string|bool File path if needle was found, false otherwise.
+	 */
+	final protected static function file_str_contains( array $files, $needle ) {
```
Why does this need to static?

felixarntz replied:
Because these methods have nothing to do with the class instance. Static methods are bad when they manage state, but all of the static methods in this class are stateless.

spacedmonkey replied:
Static methods are pointless. Why not just have a namespaced function? Is there is no benefit for these being static, then make them not static.

felixarntz replied:
I don't think that's correct. If static methods are pointless, why do they exist in lots of programming languages? One benefit of using a static method here instead of a function is that it allows the method to be scoped to a specific class (`protected`) rather than a function any code can call. The more code is public, the harder it is to make changes without breaking backward compatibility.

spacedmonkey commented on includes/Checker/Checks/Abstract_File_Check.php:
```diff
@@ -0,0 +1,227 @@
+	 * @param string $extension File extension to match.
+	 * @return array Filtered $files list.
+	 */
+	final protected static function filter_files_by_extension( array $files, $extension ) {
```
Why does this need to be static?

felixarntz replied:
See my reply in https://github.com/10up/plugin-check/pull/175#discussion_r1205834776.

spacedmonkey commented on includes/Checker/Checks/Abstract_File_Check.php:
```diff
@@ -0,0 +1,227 @@
+	 * @param array $extensions List of file extensions to match.
+	 * @return array Filtered $files list.
+	 */
+	final protected static function filter_files_by_extensions( array $files, array $extensions ) {
```
Why does this need to be static?

felixarntz replied:
Same as https://github.com/10up/plugin-check/pull/175#discussion_r1205834776

spacedmonkey commented on includes/Checker/Checks/File_Type_Check.php:
```diff
@@ -0,0 +1,184 @@
+			foreach ( $hidden_files as $file ) {
+				$result->add_message(
+					true,
+					'Hidden files are not permitted.',
```
Translation?

spacedmonkey commented:
<img width="1594" alt="Screenshot 2023-05-25 at 12 52 26" src="https://github.com/10up/plugin-check/assets/237508/b3f2e9d4-c18f-43ed-9618-e31d9a65eb03">

Styling of this error looks a little strange.

spacedmonkey commented on includes/Checker/Checks/Abstract_File_Check.php:
```diff
@@ -0,0 +1,227 @@
+	final protected static function file_str_contains( array $files, $needle ) {
+		// Backward compatibility prior to PHP 8.
+		$find_needle = 'str_contains';
+		if ( ! function_exists( 'str_contains' ) ) {
```
Why would this need exist?

felixarntz replied:
See https://github.com/10up/plugin-check/pull/175#discussion_r1205830647, will fix.

Commit:
Avoid unnecessary BC shim for str_contains.

Commit:
Avoid repeated code to amend result with file errors.

Commit:
Translate all new error messages.

felixarntz commented:
@mukeshpanchal27 @jjgrainger @spacedmonkey Please take another look, I have addressed all your feedback.

There is one PHPStan failure, however that is invalid since `str_contains` is available in all WordPress versions that we support. How can we instruct PHPStan to ignore that? cc @swissspidy 
```
Error: Function str_contains not found.
 ------ --------------------------------------------------------------------- 
  Line   includes/Checker/Checks/Abstract_File_Check.php                      
 ------ --------------------------------------------------------------------- 
  158    Function str_contains not found.                                     
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols   
 ------ ---------------------------------------------------------------------
```

swissspidy committed:
Ignore PHPStan error for now

swissspidy commented:
> There is one PHPStan failure, however that is invalid since `str_contains` is available in all WordPress versions that we support. How can we instruct PHPStan to ignore that? cc @swissspidy

Great question!

Looks like the stubs don't include that polyfill, maybe that's a bug.

For now I recommend just ignoring this error, see 502400ee15d4ed70524980c2cb8a9a88729bfea6

felixarntz commented:
Thanks @swissspidy! 🙌

spacedmonkey commented on includes/Checker/Checks/Abstract_File_Check.php:
```diff
@@ -0,0 +1,219 @@
+	 * @param string $file The file name.
+	 * @return string The file contents.
+	 */
+	private static function file_get_contents( $file ) {
+		if ( isset( self::$file_contents_cache[ $file ] ) ) {
+			return self::$file_contents_cache[ $file ];
+		}
+
+		self::$file_contents_cache[ $file ] = file_get_contents( $file );
+
+		return self::$file_contents_cache[ $file ];
+	}
```
So we are using this as a cache?

felixarntz replied:
Yes, this is particularly relevant as the same file contents may be needed again across multiple checks in a single request. Note that this was inspired by https://github.com/WordPress/plugin-check/blob/trunk/checks/check-base.php#L159 which does the same thing, which makes sense to me.

mukeshpanchal27 reviewed (approved):
Thanks @felixarntz for the update. The changes look good to me.

jjgrainger reviewed (approved):
Thanks @felixarntz changes look good, approved
